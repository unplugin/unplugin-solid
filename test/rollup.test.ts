import process from "node:process";

import { rollupBuild, testFixtures } from "@sxzz/test-utils";
import Oxc from "unplugin-oxc/rollup";
import ViteSolid from "vite-plugin-solid";
import { describe, expect } from "vitest";

import Solid from "../src/rollup";
import type { Options } from "../src/types";

async function getCode(file: string, plugin: any) {
	const bundle = await rollupBuild(file, [plugin, Oxc()], {
		external: ["solid-js"],
	});

	return bundle.snapshot;
}

function createPlugins(opt: Options & { root: string }) {
	const vite = ViteSolid(opt);
	// @ts-expect-error
	vite.configResolved!({
		root: opt.root,
		command: "build",
		dev: opt.dev,
		build: {
			sourcemap: false,
		},
		define: {},
		logger: {},
	} as any);

	return {
		unplugin: Solid(opt),
		vite,
	};
}

describe("rollup", async () => {
	await testFixtures(
		"test/fixtures/*.{js,ts,jsx,tsx}",
		async (args, id) => {
			const { unplugin, vite } = createPlugins({
				root: process.cwd(),
				dev: args.dev,
			});

			const viteCode = await getCode(id, vite);
			const unpluginCode = await getCode(id, unplugin);

			expect(viteCode).toBe(unpluginCode);

			return unpluginCode.replaceAll(
				/(["']__file["']\s*,\s*['"]).*?(['"])/g,
				(_, s1, s2) => `${s1}#FILE#${s2}`,
			);
		},
		{
			params: [["dev", [true, false]]],
			promise: true,
		},
	);
});
