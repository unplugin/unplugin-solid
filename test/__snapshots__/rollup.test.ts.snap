// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rollup > test/fixtures/basic.tsx > dev = false 1`] = `
"// basic.js
import { template } from 'solid-js/web';

var _tmpl$ = /* @__PURE__ */ template(\`<div>Hello World\`);
const Foo = _tmpl$();

export { Foo as default };
"
`;

exports[`rollup > test/fixtures/basic.tsx > dev = true 1`] = `
"// basic.js
import { template } from 'solid-js/web';

var _tmpl$ = /* @__PURE__ */ template(\`<div>Hello World\`);
const Foo = _tmpl$();

export { Foo as default };
"
`;

exports[`rollup > test/fixtures/syntax.tsx > dev = false 1`] = `
"// syntax.js
import { template, createComponent, delegateEvents } from 'solid-js/web';
import { createSignal, onMount, onCleanup, createEffect, Suspense } from 'solid-js';

var _tmpl$ = /* @__PURE__ */ template(\`<button>+\`);
function Foo(props) {
	const [count, setCount] = createSignal(0);
	onMount(() => {
		console.log("mounted");
		onCleanup(() => {
			console.log("unmounted");
		});
	});
	createEffect(() => {
		console.log("count:", count());
	});
	return [
		"Children:",
		createComponent(Suspense, {
			fallback: "loading...",
			get children() {
				return props.children;
			}
		}),
		(() => {
			var _el$ = _tmpl$();
			_el$.$$click = () => setCount(count() + 1);
			return _el$;
		})()
	];
}
delegateEvents(["click"]);

export { Foo as default };
"
`;

exports[`rollup > test/fixtures/syntax.tsx > dev = true 1`] = `
"// syntax.js
import { template, createComponent, delegateEvents } from 'solid-js/web';
import { createSignal, onMount, onCleanup, createEffect, Suspense } from 'solid-js';

var _tmpl$ = /* @__PURE__ */ template(\`<button>+\`);
function Foo(props) {
	const [count, setCount] = createSignal(0);
	onMount(() => {
		console.log("mounted");
		onCleanup(() => {
			console.log("unmounted");
		});
	});
	createEffect(() => {
		console.log("count:", count());
	});
	return [
		"Children:",
		createComponent(Suspense, {
			fallback: "loading...",
			get children() {
				return props.children;
			}
		}),
		(() => {
			var _el$ = _tmpl$();
			_el$.$$click = () => setCount(count() + 1);
			return _el$;
		})()
	];
}
delegateEvents(["click"]);

export { Foo as default };
"
`;
